File: FlinkSZ1128/src/main/scala/com/atguigu/day4/SideOutputExample.scala
package com.atguigu.day4

import com.atguigu.day2.{SensorReading, SensorSource}
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

object SideOutputExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream = env
      .addSource(new SensorSource)
      .process(new FreezingMonitor)

    stream
      .getSideOutput(new OutputTag[String]("freezing-alarms"))
      .print()
    stream.print() // 打印主流
    env.execute()
  }

  // 为什么用`ProcessFunction`? 因为没有keyBy分流
  class FreezingMonitor extends ProcessFunction[SensorReading, SensorReading] {
    // 定义侧输出标签
    lazy val freezingAlarmOutput = new OutputTag[String]("freezing-alarms")

    // 来一条数据，调用一次
    override def processElement(value: SensorReading, ctx: ProcessFunction[SensorReading, SensorReading]#Context, out: Collector[SensorReading]): Unit = {
      if (value.temperature < 32.0) {
        // 将报警信息发送到侧输出流
        ctx.output(freezingAlarmOutput, s"传感器ID为 ${value.id} 的传感器发出低于32华氏度的低温报警！")
      }
      out.collect(value) // 在主流上，将数据继续向下发送
    }
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day04/SideOutputExample.java
package com.atguigu.day04;

import com.atguigu.day02.util.SensorReading;
import com.atguigu.day02.util.SensorSource;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;

public class SideOutputExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<SensorReading> stream = env.addSource(new SensorSource());

        OutputTag<String> outputTag = new OutputTag<String>("output") {};
        OutputTag<String> outputTag1 = new OutputTag<String>("output1") {
        };

        SingleOutputStreamOperator<SensorReading> process = stream
                .process(new ProcessFunction<SensorReading, SensorReading>() {
                    @Override
                    public void processElement(SensorReading sensorReading, Context context, Collector<SensorReading> collector) throws Exception {
                        if (sensorReading.temperature < 32.0) {
                            context.output(outputTag, "小于32度的温度值发送到侧输出流中！");
                        }
                        if (sensorReading.temperature > 100.0) {
                            context.output(outputTag1, "larger than 100.0");
                        }
                        collector.collect(sensorReading);
                    }
                });

        SingleOutputStreamOperator<SensorReading> process1 = stream
                .keyBy(r -> r.id)
                .process(new KeyedProcessFunction<String, SensorReading, SensorReading>() {
                    @Override
                    public void processElement(SensorReading sensorReading, Context context, Collector<SensorReading> collector) throws Exception {
                        if (sensorReading.temperature < 32.0) {
                            context.output(outputTag, "小于32度的温度值发送到侧输出流中！");
                        }
                        collector.collect(sensorReading);
                    }
                });

        process1.getSideOutput(outputTag).print();


        process.getSideOutput(outputTag).print();
        process.getSideOutput(outputTag1).print();

        env.execute();
    }
}
